python/pyspark/sql/types.py

```diff
@@ -64,6 +64,8 @@ def json(self):
                           sort_keys=True)
 
 
+# This singleton pattern does not work with pickle, you will get
```

Comment:
Would it be possible to have singletons that pickled properly if we implemented a custom `__reduce__()` method?  Was the old `__eq__` behavior an important performance optimization?